### PR TITLE
[de] compound improvements

### PR DESCRIPTION
**Pull request by affemitkaraffe**
- refactoring
- make sure that compound parts are either nouns, verb stems or other prefixes are spelled correctly
- make SUB:NOM:SIN and no infix s the default for compounds parts
- allow gender neutral forms in compounds (e. g. SEO-Expert*innen)
- do not allow old spelling in German compounds (e. g. Abflußleitung)
- allow for more special prefixes in German compounds (e. g. Thermal-, Anti-, Ur-)
- distinguish between nouns that require singular/plural (e. g. Gasehersteller, Hundzwinger)

-problems:
- compounds that pass the checks but are nonsensical (e. g. Eisbergspritze) will lead to new FNs
- but AI rules are expected to detect between 66% - 75% of those FNs

### CODE REVIEW

File: languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java [hunk 1]
```diff
@@ -2270,83 +2311,161 @@ protected boolean ignorePotentiallyMisspelledWord(String word) throws IOExceptio
...
-    }
     return false;
   }
 
+  private boolean isValidGenderNeutralWord(List<String> parts, String word) {
+    int start = 0;
+    for (String part : parts) {
+      int end = start + part.length();
+
+      String to_check = word.substring(start, end);
```
Comment (danielnaber) [inline languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java:2418]: please use CamelCase (`toCheck`)

File: languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java [hunk 2]
```diff
@@ -2270,83 +2311,161 @@ protected boolean ignorePotentiallyMisspelledWord(String word) throws IOExceptio
...
+      if (to_check.startsWith("I") && start > 0) {
+        // e. g. AktienIndex
+        return false;
+      }
+      if (to_check.matches(".*((?<=(\\w))In).*")) {
```
Comment (danielnaber) [inline languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java:2423]: Please pre-compile all regex (as with `private static final Pattern VERBAND_COMP....`)

File: languagetool-language-modules/de/src/main/java/org/languagetool/tokenizers/de/GermanCompoundTokenizer.java [hunk 3]
```diff
@@ -202,6 +202,39 @@ public GermanCompoundTokenizer(boolean strictMode) throws IOException {
     wordSplitter.addException("Festigung", asList("Festigung"));
     wordSplitter.addException("Allerheiligen", asList("Allerheiligen"));
     wordSplitter.addException("Druckerpressen", asList("Drucker", "pressen"));
+    wordSplitter.addException("Habitat", asList("Habitat"));
```
Comment (danielnaber) [inline languagetool-language-modules/de/src/main/java/org/languagetool/tokenizers/de/GermanCompoundTokenizer.java:205]: This is unrelated, isn't it? Could be its own PR for easier review.

File: languagetool-language-modules/de/src/main/java/org/languagetool/tokenizers/de/GermanCompoundTokenizer.java [hunk 4]
```diff
@@ -202,6 +202,39 @@ public GermanCompoundTokenizer(boolean strictMode) throws IOException {
...
+    wordSplitter.addException("Augarten", asList("Augarten"));
+    wordSplitter.addException("Auszeit", asList("Auszeit"));
+    wordSplitter.addException("Bewegtbild", asList("Bewegtbild"));
```
Comment (danielnaber) [inline languagetool-language-modules/de/src/main/java/org/languagetool/tokenizers/de/GermanCompoundTokenizer.java:208]: Why not split into `Bewegt` + `Bild`?

File: languagetool-language-modules/de/src/test/java/org/languagetool/rules/de/GermanSpellerRuleTest.java [hunk 5]
```diff
@@ -134,12 +185,137 @@ public void testIgnoreMisspelledWord() throws IOException {
...
     // special cases:
     assertFalse(rule.ignorePotentiallyMisspelledWord("Actionsspaß"));
-    assertFalse(rule.ignorePotentiallyMisspelledWord("Jungsnamen"));
-    assertFalse(rule.ignorePotentiallyMisspelledWord("Aufschwungsphase"));
-    assertFalse(rule.ignorePotentiallyMisspelledWord("Absprungsrate"));
+    //assertTrue(rule.ignorePotentiallyMisspelledWord("Jungsnamen")); // 'Jungs' is colloquial for 'Jungen'
+    assertFalse(rule.ignorePotentiallyMisspelledWord("Aufschwungsphase")); // see prohibit.txt
+    assertFalse(rule.ignorePotentiallyMisspelledWord("Absprungsrate")); // see prohibit.txt
+    //
+    assertFalse(rule.ignorePotentiallyMisspelledWord("Ahornswald"));
```
Comment (danielnaber) [inline languagetool-language-modules/de/src/test/java/org/languagetool/rules/de/GermanSpellerRuleTest.java:219]: Is there any use in adding this many tests? Does it actually test a different code path, or does it always test the same code?

**Review by danielnaber — CHANGES_REQUESTED** [review]

File: languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java [hunk 6]
```diff
@@ -136,12 +136,14 @@ public class GermanSpellerRule extends CompoundAwareHunspellRule {
...
...ähe|kraut|krippe|kuriosität|kurve|kusine|küste|laie|laterne|laute|legende|lehne|leise|lerche|leuchte|lippe|loge|lotse|länge|läuse|löwe|lücke|luke|made|maske|maßnahme|matriarchin|menge|mensch|metapher|methode|metropole|miene|miete|million|mineral|miniatur|mitte|maus|monarch(in)?|mormone|mücke|mühle|musikant(in)?|mysterium|nerv|niederlage|nixe|nonne|note|obdachlose|ode|organist|panne|parzelle|pate|patient|patriarch(in)?|petze|pfanne|pfeife|platte|polle|pomade|pomeranze|posse|praktikant(in)?|prise|prominente|prototyp|prälat|puppe|pädophile|rabe|radikale|rakete|rampe|ranke|rate|raupe|rendite|repressalie|rest|riese|rinde|rind|robbe|robe|romanist|rose|ross|route|nummer|runde|röhre|rübe|salbe|schabe|schale|scheide|schelle|schenke|schere|sphäre|dicke|kröte|schlampe|schlange|schluchte|schmiere|schnake|schnalle|schneide|schnelle|schokolade|schotte|schurke|schwabe|schwalbe|schwede|schwule|seele|seide|seite|serie|silbe|sonne|sorge|sorte|spanne|sparte|sperre|spitze|sproße|spule|steppe|straße|streife|studie|stunde|stütze|tabelle|tinte|tote|toilette|torte|traube|treffe|treppe|truhe|träne|tunte|tüte|urne|vene|versicherte|verwandte|virtuose|vorname|waffe|wanne|ware|watte|wehe|welle|wiese|zentrum");
```
Comment (Luisa-LT) [inline languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java:140]: There are some words where the plural form would not (always) be right for compounds, e.g., _miniatur_ (_Miniaturausgabe, Mi­ni­a­tur­bild, Mi­ni­a­tur­ge­mäl­de, Miniaturgolf_ etc.) and also words that would rather be shortened than used in plural in compounds, like _leuchte -> leucht (Leuchtmittel), metropole -> metropol (Metropolregion), mineral -> mineral (Mineralwasser)_, _radikal -> radikal (Radikaldiät)_ etc.

File: languagetool-language-modules/de/src/main/resources/org/languagetool/resource/de/other_prefixes.txt [hunk 7]
```diff
@@ -0,0 +1,411 @@
+#list of prefixes comes from https://github.com/danielnaber/jwordsplitter/
+sach
+hilf
+eß
```
Comment (Luisa-LT) [inline languagetool-language-modules/de/src/main/resources/org/languagetool/resource/de/other_prefixes.txt:4]: "eß" should be removed as words like "Eßgewohnheiten" would be accepted

File: languagetool-language-modules/de/src/main/resources/org/languagetool/resource/de/other_prefixes.txt [hunk 8]
```diff
@@ -0,0 +1,411 @@
+#list of prefixes comes from https://github.com/danielnaber/jwordsplitter/
+sach
+hilf
```
Comment (Luisa-LT) [inline languagetool-language-modules/de/src/main/resources/org/languagetool/resource/de/other_prefixes.txt:3]: "hilf" would need an "s" (_Hilfsgüter_)

File: languagetool-language-modules/de/src/main/resources/org/languagetool/resource/de/other_prefixes.txt [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+eß
+sprach
+sozial
+freß
```
Comment (Luisa-LT) [inline languagetool-language-modules/de/src/main/resources/org/languagetool/resource/de/other_prefixes.txt:7]: same like "eß" -> should be removed as we would accept words like _Freßfeinde_ (old spelling)

**Review by Luisa-LT — CHANGES_REQUESTED** [review]

File: languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java (hunk 6 above)
Comment (St-ac-y) [reply languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java:140]: yes, similar to `adresse` -> Adressbuch etc.

File: languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java (hunk 1 above)
Comment (affemitkaraffe) [reply languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java:2418]: I don't think I can just check for CamelCase here. Since the Binnen-I can be used for gendering, I need to check the individual parts of the compound as well ('SEO-ExpertInnen' VS 'SEO-ExpertInnern')

File: languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java (hunk 6 above)
Comment (affemitkaraffe) [reply languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java:140]: >There are some words where the plural form would not (always) be right for compounds

There are a lot of such words in German. It is extremely time-consuming to collect an exhaustive list for those cases. That's why I decided to go for either 'always plural' or 'always singular' for less common cases. If you think 'This word is more often singular than plural when used in compounds' or the other way around, I can change the 'default numerus'. 

But creating lists that differantiate between singular and plural for words (e. g. Figurproblem vs Figurengedicht), is really time-consuming. I did this for common words that sometimes require infix s and sometime don't allow it, but this becomes a Sisyphean task at some point.

> also words that would rather be shortened than used in plural in compounds
Words like 'Leuchtmittel' are already accepted by other parts of the code. But we should also allow words like 'Leuchtenhersteller', but we should still flag typos like 'Leuchtemittel' or 'Leuchtehersteller'. This is what the code aims at.

File: languagetool-language-modules/de/src/main/resources/org/languagetool/resource/de/other_prefixes.txt (hunk 9 above)
Comment (affemitkaraffe) [reply languagetool-language-modules/de/src/main/resources/org/languagetool/resource/de/other_prefixes.txt]: I think I had an old version of the german_prefix file

File: languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java (hunk 1 above)
Comment (danielnaber) [reply languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java:2422]: I meant the variable name, it should be `toCheck` instead of `to_check`, according to Java conventions.

**Review by danielnaber — CHANGES_REQUESTED** [review]

File: languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java [hunk 10]
```diff
@@ -2363,6 +2486,341 @@ private boolean isOnlyPluralNoun(String word) throws IOException {
...
+    if (isNounNomSin(part1) && isVerbStem(part2) && isNoun(compound2)) {
+      // e.g. "Wein" + "kühl" + "schrank"
+      return true;
+    }
+    return false;
+  }
+
+  private boolean checkInfixSForPart1Part2Combination(String part1, String part2) throws IOException {
+    // For some part1-and-part2 combinations an infix s is correct or incorrect
+    String part2_lemma = findLemmaForNoun(removeTrailingHyphen(part2));
```
Comment (danielnaber) [inline languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java:2599]: camelCase: `part2Lemma`

File: languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java [hunk 11]
```diff
@@ -2363,6 +2486,341 @@ private boolean isOnlyPluralNoun(String word) throws IOException {
...
+
+  private boolean isProbablyTypo(String word) {
+    return COMPOUND_TYPOS.matcher(word).matches() || COMPOUND_END_TYPOS.matcher(word).matches();
+  }
+
+  private boolean isValidCamelCase(String input) {
+    // Check if the string contains any instance of camel casing
+    // TODO Is this already implemented in StringTools::isCamelCase?
+
+    return !input.matches(".*(\\p{Ll}\\p{Lu}|\\p{Lu}{2,}\\p{Ll}).*");
```
Comment (danielnaber) [inline languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java:2687]: pre-compile

File: languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java [hunk 12]
```diff
@@ -2363,6 +2486,341 @@ private boolean isOnlyPluralNoun(String word) throws IOException {
...
+        tokensWithHyphens.add(token);
+        currentPos += token.length();
+      }
+      return tokensWithHyphens;
+    }
+
+  private List<String> avoidInfixSAsSingleToken(List<String> parts) {
+    // If a part equals "s", append it to its predecessor
+    //   example: "Priorität", "s", "ding" -> "Prioritäts", "ding"
+    List<String> fixed_parts = new ArrayList<String>();
```
Comment (danielnaber) [inline languagetool-language-modules/de/src/main/java/org/languagetool/rules/de/GermanSpellerRule.java:2722]: camelCase

**Review by danielnaber — CHANGES_REQUESTED** [review]

**Comment by Luisa-LT** [comment]
🚀 🚀 🚀

**Comment by maphjo** [comment]
:tada: :rocket: :star2:
